charger/eebus.go

```diff
@@ -311,6 +316,128 @@ func (c *EEBus) optimizationSelfConsumptionAvailable() bool {
+	data, err := c.cc.GetData()
+	if err != nil {
+		return err
+	}
+
+	var chargingPlan communication.EVChargingPlan
+
+	tariffGrid := 0.30
+	tariffFeedIn := 0.10
+	maxPower := c.lp.GetMaxPower()
+
+	switch data.EVData.ChargingStrategy {
+	case communication.EVChargingStrategyEnumTypeNoDemand, communication.EVChargingStrategyEnumTypeUnknown:
+		// The EV has no power demand or we don't know it yet, so we shouldn't get here
+		// TODO: why did we get here?
+
+		// lets do 24 1 hour slots with maximum power, power will be adjusted via Overload Protection limits
+		for i := 0; i < 24; i++ {
+			chargingPlan.Slots = append(chargingPlan.Slots, communication.EVChargingSlot{
+				Duration: time.Duration(1) * time.Hour,
+				MaxValue: maxPower,
+				Pricing:  tariffGrid,
+			})
+		}
+		chargingPlan.Duration = time.Duration(24) * time.Hour
+	case communication.EVChargingStrategyEnumTypeDirectCharging:
+		// The EV is in direct charging mode
+
+		// Does it support self consumption?
+		if c.optimizationSelfConsumptionAvailable() && !c.forcePVLimits {
+			// this should mean that any mode in evcc is ignored and the EV is in full control
+			// TODO: is this the right approach?
+
+			// lets do one 24 hour slot with maximum power, power will be adjusted via Overload Protection limits
+			chargingPlan.Slots = append(chargingPlan.Slots, communication.EVChargingSlot{
+				Duration: time.Duration(24) * time.Hour,
+				MaxValue: maxPower,
+				Pricing:  tariffGrid,
+			})
+			chargingPlan.Duration = time.Duration(24) * time.Hour
+		} else {
+			// in this mode we need to enforce the evcc modes
+
+			// we need to create a 24h charging plan
+			chargingPlan.Duration = time.Duration(24) * time.Hour
+
+			currentMode := c.lp.GetMode()
+			switch currentMode {
+			case api.ModeNow, api.ModeMinPV:
+				// lets do one 24 hour slot with maximum power, power will be adjusted via Overload Protection limits
+				chargingPlan.Slots = append(chargingPlan.Slots, communication.EVChargingSlot{
+					Duration: time.Duration(24) * time.Hour,
+					MaxValue: maxPower,
+					Pricing:  tariffGrid,
+				})
+				chargingPlan.Duration = time.Duration(24) * time.Hour
+			case api.ModePV:
+				// lets do 24 1 hour slots with maximum power, power will be adjusted via Overload Protection limits
+				// but set the nightly hours to 0 W, we assume those to be from 20:00 to 07:00
+				now := time.Now()
+				for i := 0; i < 24; i++ {
+					power := maxPower
+					pricing := tariffFeedIn
+					if now.Hour()+i >= 20 || now.Hour()+i < 7 {
+						power = 0.0
+						pricing = tariffGrid
+					}
+					chargingPlan.Slots = append(chargingPlan.Slots, communication.EVChargingSlot{
+						Duration: time.Duration(1) * time.Hour,
+						MaxValue: power,
+						Pricing:  pricing,
+					})
+				}
+				chargingPlan.Duration = time.Duration(24) * time.Hour
+			case api.ModeOff:
+				// lets do 24 1 hour slots with 0 W, so it wakes at once an hour to check back
+				for i := 0; i < 24; i++ {
```

Comment:
sollte man den Loop nicht mal in eine Funktion verpacken?

Reply:
Momentan noch nicht. Das wird komplexer wenn ich die Tarifwerte von evcc bekomme. Dann mal schauen.